R/datasets.R

```diff
@@ -75,193 +63,201 @@ read_table_arrow <- function(dataset, branch = NULL, transaction = NULL, timeout
-    }
-    path_ <- paste(target, file_path, sep = "/")
-    if (!dir.exists(dirname(path_))) {
-      dir.create(dirname(path_), recursive = TRUE)
+
+  datasets <- get_datasets_client()
+  create_parent_and_download <- function(file_path) {
+    path <- file.path(target, gsub("/", .Platform$file.sep, file_path))
+    if (!dir.exists(dirname(path))) {
+      dir.create(dirname(path), recursive = TRUE)
     }
-    download_file(dataset, file_path, path_, ...)
+    file <- datasets$get_file_content(dataset$rid, file_path, branch_id = dataset$branch,
+                                      start_transaction_rid = dataset$start_transaction_rid,
+                                      end_transaction_rid = dataset$end_transaction_rid)
+    writeBin(file$content, path)
+    return(path)
   }
-  cat(paste0("Downloading ", length(files), " files to local directory ", target, "\n"))
-  lapply(files, create_parent_and_download)
-  invisible(NULL)
+
+  return(sapply(files, create_parent_and_download))
 }
 
 #' Upload a local file or folder to a Foundry Dataset.
 #'
-#' @param local_file A character string representing the location of the file or folder to upload.
+#' @param files The local files and folders to upload.
 #' If a folder is provided, all files found recursively in subfolders will be uploaded.
-#' @param dataset A Dataset or a character string representing the RID or dataset path.
-#' @param branch The dataset branch.
-#' @param txn_type The type of the transaction, must be one of 'UPDATE', 'APPEND', 'SNAPSHOT'.
+#' @param alias The alias representing the Dataset.
+#'
+#' @return The list of local file paths and corresponding file name in the Foundry dataset.
 #'
 #' @export
-upload_file <- function(local_file, dataset, branch = NULL, txn_type = "UPDATE") {
-  if (!file.exists(local_file)) {
-    stop("The file to upload does not exist: ", local_file)
+datasets.upload_files <- function(files, alias) { # nolint: object_name_linter
+  dataset <- get_alias(alias)
+  if (dataset$type != "dataset") {
+    stop(sprintf("Expected alias '%s' to be a dataset, found %s", alias, dataset$type))
   }
-  if (!txn_type %in% c("UPDATE", "APPEND", "SNAPSHOT")) {
-    stop("The transaction type must be one of 'UPDATE', 'APPEND', 'SNAPSHOT'")
+
+  missing_files <- sapply(files, function(x) !file.exists(x))
+  if (length(missing_files) > 0) {
+    stop(sprintf("The following local files do not exist: %s", paste(missing_files, collapse = ", ")))
+  }
+
+  get_file_to_upload <- function(local_file) {
+    if (!dir.exists(local_file)) {
+      return(list(list(file_name = basename(local_file), file = local_file)))
+    }
+    return(lapply(list.files(local_file, recursive = TRUE), function(path) {
+        list(file_name = gsub(.Platform$file.sep, "/", path), file = file.path(local_file, path))
+      }))
+  }
+
+  files_to_upload <- unlist(lapply(files, get_file_to_upload), recursive = FALSE)
+  files_to_upload <- files_to_upload[!duplicated(files_to_upload)]
```

Comment:
If the same file is requested to be uploaded twice, we just do the upload once. It could happen if you provide a file as well as its parent directory.